### PR TITLE
fetch Quote Tweets

### PR DESCRIPTION
**Pull request by kylewm**
Adds fetch_quotes kwarg to get_activities() to include quote tweets
in a post's tags field. They are represented just like retweets but
use the (made up) verb "quote" instead of "share".

Uses the Twitter search API to search for references to other tweets
by ID, in groups of 5.

### CODE REVIEW

**Comment by kylewm** [comment]
I guess I'm pretty uncertain how to represent quote tweets. I don't *think* it's correct to give them a verb "quote", which the code above does ... the verb should really just be "post", with the quoted tweet in the list of tags.

But then how do you tell which tags are quotes, vs just regular links or other stuff

**Comment by snarfed** [comment]
oh man, cool! thank you for doing this!

i assume you're asking about how to represent these in AS? i don't feel strongly: either quote or post with tag distinguished somehow is fine with me.

the discussion in #504 on how to represent them in mf2 is maybe more important. you probably already saw our tentative conclusion to do simple link mention, since that's what the raw tweet text itself does. (quoted tweet url in content, usually at end.) maybe wrapped in h-cite of the quoted tweet's full content to mirror Twitter's rendering style?

**Comment by kylewm** [comment]
Yeah, how to represent in AS. Do you think we'll need a new Bridgy endpoint for quotes, or just serve them like mentions?

**Comment by snarfed** [comment]
nah, i think the `/post/...` endpoint like for mentions will work. i wonder if we should even fold quote tweets into `fetch_mentions` instead of adding the new kwarg? we kind of already have kwarg proliferation as it is :P,  but still maybe worth resisting more...?

I'm also thinking about the additional load. we fetch the last 50 tweets (ignoring etag in granary), so this adds 10 api requests per poll? reasonable, but noticeable. https://dev.twitter.com/rest/public/search says:

> Limit your searches to 10 keywords and operators.

so maybe bump up to 10 per call?

**Comment by kylewm** [comment]
>  so maybe bump up to 10 per call?

heh, depends how you interpret "10 keywords and operators". I was thinking 5 because 5 ids + 4 ORs = 9 keywords and operators.

**Comment by kylewm** [comment]
>  i wonder if we should even fold quote tweets into fetch_mentions instead of adding the new kwarg?

I think this makes good sense, especially since there don't seem to be analogues for quote tweets on any other silo.

**Comment by kylewm** [comment]
So empirically batches of up to 20 seems to succeed (can't confrm that it actually accurately finds all 20 IDs). A batch of 25 errors with "Missing or invalid url parameter" I would've expected "Sorry, your query is too complex" based on the documentation but oh well.

**Comment by snarfed** [comment]
either 10 or 20 sgtm! also see https://github.com/snarfed/granary/blob/04f39a4f9a3a068e23d73886ce97e07856a8c760/granary/facebook.py#L373 for an example of another place where we batch things into API calls with a cap.

bigger question: an alternative would be to return quote tweets as top level activities, instead of tags, like we currently do for mentions. (granted, we don't have anywhere else to put those.) this would make the `get_activities` code simpler, but let's see...it would make OPD way harder? i obviously haven't thought it through; just curious. :P

also, interesting minor tradeoff: we could search for full tweet urls instead of just ids, which would cut down false positives (probably very rare anyway :P), at the cost of probably fewer per API call. ids only is probably the right call, but maybe worth a comment? meh.

**Comment by kylewm** [comment]
false positives are interesting, I noticed earlier that [searching for a tweet with a photo](https://twitter.com/search?f=tweets&q=678304879572480000&src=typd) includes the original itself.

Searching by full URL works but it does seem to be a little more sensitive. e.g. [full url](https://twitter.com/search?q=https%3A%2F%2Ftwitter.com%2Fxychelsea%2Fstatus%2F674981556591706113&src=typd) returns way fewer results (4) than [URL without scheme](https://twitter.com/search?f=tweets&q=twitter.com%2Fxychelsea%2Fstatus%2F674981556591706113&src=typd) or [ID only](https://twitter.com/search?f=tweets&q=674981556591706113&src=typd) (which both return 12)

**Comment by kylewm** [comment]
> an alternative would be to return quote tweets as top level activities

Well I guess that is more in line with how Twitter presents them (i.e. not as threaded replies, but as separate tweets in someone else's timeline). And since we're including quotes with mentions when fetch_mentions=True, I like that we'd treat them similarly.

Hmm this will add a little to Step 2 of Bridgy's tasks.poll -- it'll need to check each activity to see if the owner of an attachment is the current user, and if so add the top-level activity to the responses to propagate (very similar to how it handles mentions or links now).

Will it add complexity to original_post_discovery? The quoted tweet will be in an attachment.. I'd think it'd be pretty similar to OPD for reply tweets, but looking at the code now I don't honestly understand how replies work. If you call original_post_discovery on the reply tweet, I'm not seeing how it finds the original post belonging to the in-reply-to tweet. Maybe will be clearer in the morning :)

**Comment by kylewm** [comment]
original post discovery is handled for comments by Bridgy's CommentHandler which gets the post_id from the handler's URL and explicitly does OPD on that; same for LikeHandler and RepostHandler.

So actually I don't think it particularly matters whether quotes are represented as top-level objects or not -- Bridgy's poll task pulls out responses and puts them all in a flat list along with top-level mentions anyway. 

I think I will need to relax the OPD code a little so that it always does OPD on attached notes/posts. If a quoted tweet has an original URL, then the quoting tweet will mention that URL.

File: granary/twitter.py [hunk 1]
```diff
@@ -291,6 +297,29 @@ def _user():
           retweet_calls += 1
           cache_updates['ATR ' + id] = count
 
+    if fetch_quotes:
+      # organize quote tweets by id of the original
+      quote_tweets = {}
```
Comment (snarfed) [inline granary/twitter.py]: very minor: this could be a `defaultdict(list)`

File: granary/twitter.py [hunk 2]
```diff
@@ -291,6 +297,29 @@ def _user():
...
+      # search for 5 ids at a time based on the guideline
+      # "Limit your searches to 10 keywords and operators."
+      # https://dev.twitter.com/rest/public/search
+      for five_tweets in [tweets[i:i + 5] for i in xrange(0, len(tweets), 5)]:
+        url = API_SEARCH_URL % {
+          'q': urllib.quote_plus(
+            ' OR '.join((t['id_str'] for t in five_tweets))),
```
Comment (snarfed) [inline granary/twitter.py]: funny, reminds me of https://github.com/snarfed/bridgy/blob/444b05710c15bdd2713e5cb8fe2f15a478d0e8cd/models.py#L273 and https://github.com/snarfed/bridgy/blob/444b05710c15bdd2713e5cb8fe2f15a478d0e8cd/twitter.py#L67 etc

**Comment by snarfed** [comment]
right, that's the change we'd need for OPD. that can be either here or in another PR.

and i'm fine with attaching quotes to the quoted tweet activity like you do here. i think you're right that the total code weight is the same either way, it just shifts between granary and bridgy. so with `fetch_mentions` more ids per search, i say merge this!

another thing to think about (beyond this PR) is that it's taken a fair amount of delicate work to get the precedence right in bridgy's poll. e.g. if a tweet links to the user's site *and* @-mentions the user, the @-mention shouldn't override the link. same if the tweet is also a reply, etc. we'll need to check that for quote tweets too. that's actually related to the errant `continue` you noticed. the logic to keep that right is pretty fragile imho. some of it depends on e.g. processing replies later than user mentions, so that an @-reply response objects replaces a user mention version of the same object.

not related to this PR specifically, just fyi. it's partly why i'd avoided working on this at all. :P

**Comment by kylewm** [comment]
> I think I will need to relax the OPD code a little so that it always does OPD on attached notes/posts. If a quoted tweet has an original URL, then the quoting tweet will mention that URL.

Ugh, OK actually it's more than this -- it needs to do recursive (depth=1) original post discovery on the attached post.

It's hard to differentiate situation 1: post has an attachment which is an article on the original site, e.g. a Facebook post with a "link" property. -- in this case the attachment's URL *is* the original post URL for the silo post.

from situation 2: post has an attachment which is another post on the same service, e.g. quote tweets -- in this case, the post *mentions* the original post URL of the attached post.

**Comment by snarfed** [comment]
yup, ugh. thanks for thinking this through.

honestly, every newly shaped backfeed feature that we've added in the last year or so has made the poll-OPD-propagate pipeline more and more complicated, and taken more and more effort to get right while not also breaking anything else and still keeping the overall code remotely understandable. it's not a great trend, and it's why i deprioritized this one in particular (snarfed/bridgy#504) and kinda set a higher bar in my mind for new backfeed extensions in general.

I'm still open to getting this one all the way through, if you're up for it! we're only going to hit more complications though, both before and after shipping it. :/ so I'm also fine with leaving it in granary, or playing it by ear. up to you.

**Comment by kylewm** [comment]
thanks, that was encouraging actually. nice to be reminded that i'm not the only one making things more complicated.

i think i found a pretty targeted way to modify the original post discovery in bridgy. we'll see what you think!

**Comment by snarfed** [comment]
:+1:

**Pull request merged**